crates/sui-node/src/lib.rs

```diff
@@ -689,6 +692,47 @@ impl SuiNode {
             .ok_or_else(|| anyhow::anyhow!("Transaction Orchestrator is not enabled in this node."))
     }
 
+    /// TEMPORARY: Checks whether the binary contains a new framework version to apply at the next
+    /// epoch boundary.  If it does, the current framework is overwritten with the new framework,
+    /// preserving its version and digest, and the node panics, so that it can be restarted with the
+    /// new framework in place.
+    pub async fn check_emergency_framework_upgrade(&self, next_epoch: u64) {
+        let Some(modules) = sui_config::framework::override_sui_framework(next_epoch) else {
+            return;
+        };
+
+        let new_package = match Object::new_package(modules, TransactionDigest::genesis()) {
+            Ok(pkg) => pkg,
+            Err(e) => {
+                error!(next_epoch, "Emergency Upgrade - failed to create new package: {e:#?}");
+                return;
+            }
+        };
+
+        let store = &self.state.database;
+        let Ok(Some(curr_package)) = store.get_object(&SUI_FRAMEWORK_OBJECT_ID) else {
+            error!(next_epoch, "Emergency Upgrade - can't find existing package");
+            return;
+        };
+
+        if curr_package == new_package {
+            info!(next_epoch, "Emergency Upgrade - picking up upgraded package");
+            return;
+        }
+
+        let package_ref = curr_package.compute_object_reference();
+        match store.insert_object_direct(package_ref, &new_package).await {
+            Ok(()) => {
+                panic!("Emergency Upgrade - at epoch {next_epoch}, please restart.")
```

Comment:
Capturing the nice Slack explanation of why this works in a comment would be useful here.